### ISSUE TITLE
HEADSUP: next release (3.5) plan

### ISSUE THREAD

**Issue by hannesm** [mirage/mirage#973]
hey, I plan to do a mirage release (3.5.0) _this week_ (i.e. before the marrakesh retreat). this way we can work in Marrakesh with the new API (e.g. finally a traceroute as MirageOS unikernel) and fix potential bugs, and don't have to pin a variety of opam packages.

Changes I plan to include:
- TCP/IP layering rework (#968, including #972) - most dependent packages are already in opam-repository
- fix regression #970 
- #971
- mirage-kv 2.0.0 #974

I'm not sure how ready the dune #969 parts are (@TheLortex), I'd appreciate to get these changes in as well (but if they're not ready, let's wait for another release cycle).

are there any concerns about the above changes/plan (e.g. @g2p mirage-block changes?)? //cc @mirage/core as usual, I'll do end-to-end tests on FreeBSD/solo5, and let the CI check compilations of opam packages and mirage-skeleton unikernels.

thanks for reading, and for your contributions! for a wider announced release, I guess the current opinion is to wait for a more developer-friendly dune-based unikernel release!?

**Comment by avsm** [comment]
This sounds great -- maybe a short blog post to announce the interface changes? We're pretty close to being able to make the next release the 'big' (in terms of tooling anyway) dune-based release, so that can form a wider announcement. But the changes in this release are most definitely worthy of a blog post since they represent a long overdue maturing of net/kv interfaces!

**Comment by avsm** [comment]
Also, I'm a little worried about xen backend bitrot, so a blog post is our ultimate regression test for the xen backend :-)

**Comment by hannesm** [comment]
@avsm yes, that is a great idea! I'll try to write one up for the 3.0.0 -- 3.5.0 changes. (last posts were about 3.0.0 release).

**Comment by samoht** [comment]
Can we announce the dune support with the 4.0 release? :-)

**Comment by avsm** [comment]
@samoht yes -- this blog post is for the interface changes since 3.0->3.5

**Comment by hannesm** [comment]
status: most is PRed to opam-repository (see #974 for detailed checklist), missing: cohttp-mirage (depends on mirage-dns being released) and crunch (see discussion/raise your voice in https://github.com/mirage/ocaml-crunch/pull/47), mirage & mirage-skeleton changes (I've drafts locally, will PR later and ensure that travis is happy compiling a whole mirage-skeleton, plus testing end-to-end on my laptop)

**Comment by hannesm** [comment]
closing, since https://github.com/ocaml/opam-repository/pull/13567 is now PR'ed to opam-repository. there are still some packages (tar-mirage, mirage-qubes-ipv4, irmin, ..) which would be nice to have releases of, but they can appear in opam-repository independently. closing this, and soon opening a PR to mirage-www with the requested blog entry.

**Comment by hannesm** [comment]
another HEADSUP /cc @mirage/core
- this has been released now \o/
- this removes on lots of backends the io_page allocations
- this may lead to out of memory exceptions (I initially thought these must be memory leaks, but they are not).

The reason for memory to disappear is that no `Gc.compact ()` is called to free up the externally-allocate bigarray memory (in contrast, io-page calls `Gc.compact ()` once if allocation fails https://github.com/mirage/io-page/blob/master/lib/io_page.ml#L42). I tried to embed such a call in `Cstruct.create_unsafe`, but without much luck. I guess https://github.com/ocaml/ocaml/pull/1738 (which will be part of 4.08) may solve the issue completely. But for the moment, I don't have a good idea what to do about this. Is there a way to tweak the Gc parameters (esp. in respect to bigarrays) already? Another heuristic I could think of (but dislike) would be to call `Gc.full_major ()` every N minutes in an asynchronous task. WDYT?
